
## Browser Attribute Fingerprinting Scripts Analysis 

There are some scripts that we can pick out by name that are doing browser attribute fingerprinting:
    
1. Scripts with hs-analytics in the script_url
2. Scripts with /akam/ in the script_url
3. The fingerprintjs scripts

This notebook has used the existing method which @birdsarah has produced in analyses/issue_34_setup_and_dask_tips.ipynb to pull out these scripts.

It calculates the percentage of each of the three scripts with respect to the total number of scripts. It also calculates the percentage when the three of the scripts are considered together.

Value_1000_only dataset is used which contains all the rows of the dataset, but truncates the value field to only keep the first 1000 characters in a column called value_1000. 


In [18]:
import dask.dataframe as dd
from dask.distributed import Client

Client()

/home/soumya/.conda/envs/overscripted/lib/python3.6/site-packages/distributed/bokeh/core.py:57: UserWarning: 
Port 8787 is already in use. 
Perhaps you already have a cluster running?
Hosting the diagnostics dashboard on a random port instead.
  warnings.warn('\n' + msg)


<Client: scheduler='tcp://127.0.0.1:45465' processes=4 cores=12>

In [43]:
df = dd.read_parquet(
    '/home/soumya/Dataset/sample_10percent_value_1000_only.parquet', engine='pyarrow',
    columns=['argument_0', 'symbol', 'location', 'script_url', 'value_1000']
)
part_0 = df.get_partition(0)
print(f'''
    Complete dataset is {len(df):,} rows.
    Partition is {len(part_0):,} rows.
    Type of dataset is {type(df)}.
    Type of part_0 is {type(part_0)}.
''')
df.compute()


    Complete dataset is 11,292,867 rows.
    Partition is 37,022 rows.
    Type of dataset is <class 'dask.dataframe.core.DataFrame'>.
    Type of part_0 is <class 'dask.dataframe.core.DataFrame'>.



argument_0                       symbol  \
0           None                  window.name   
1           None                  window.name   
2           None   window.navigator.userAgent   
3           None   window.navigator.userAgent   
4           None       window.document.cookie   
5           None                  window.name   
6           None                  window.name   
7           None       window.document.cookie   
8           None       window.document.cookie   
9           None       window.document.cookie   
10          None       window.document.cookie   
11          None       window.document.cookie   
12          None       window.document.cookie   
13          None       window.document.cookie   
14          None       window.document.cookie   
15          None       window.document.cookie   
16          None       window.document.cookie   
17          None       window.document.cookie   
18          None       window.document.cookie   
19          None       window.document.cookie   
20          None       window.document.cookie   
21          None       window.document.cookie   
22          None       window.document.cookie   
23          None       window.document.cookie   
24          None       window.document.cookie   
25          None       window.document.cookie   
26          None       window.document.cookie   
27          None       window.document.cookie   
28          None       window.document.cookie   
29          None       window.document.cookie   
...          ...                          ...   
37790       None       window.document.cookie   
37791       None       window.document.cookie   
37792       None       window.document.cookie   
37793       None       window.document.cookie   
37794       None       window.document.cookie   
37795       None       window.document.cookie   
37796       None       window.document.cookie   
37797       None       window.document.cookie   
37798       None       window.document.cookie   
37799       None       window.document.cookie   
37800       None       window.document.cookie   
37801       None       window.document.cookie   
37802       None       window.document.cookie   
37803       None       window.document.cookie   
37804       None       window.document.cookie   
37805       None       window.document.cookie   
37806       None       window.document.cookie   
37807       None       window.document.cookie   
37808       None       window.document.cookie   
37809       None       window.document.cookie   
37810       None       window.document.cookie   
37811       None       window.document.cookie   
37812       None   window.navigator.userAgent   
37813       None  window.navigator.appVersion   
37814       None          window.localStorage   
37815       None     window.screen.colorDepth   
37816       None     window.screen.colorDepth   
37817       None   window.navigator.userAgent   
37818       None          window.localStorage   
37819       None   window.navigator.userAgent   

                                                location  \
0      https://vk.com/widget_comments.php?app=2297596...   
1      https://vk.com/widget_comments.php?app=2297596...   
2      https://vk.com/widget_comments.php?app=2297596...   
3      https://pos.baidu.com/s?hei=70&wid=670&di=u313...   
4      http://serienjunkies.org/smilf/smilf-season-1-...   
5      http://serienjunkies.org/smilf/smilf-season-1-...   
6      http://serienjunkies.org/smilf/smilf-season-1-...   
7                 https://www.canada.ca/en/services.html   
8                 https://www.canada.ca/en/services.html   
9                 https://www.canada.ca/en/services.html   
10                https://www.canada.ca/en/services.html   
11                https://www.canada.ca/en/services.html   
12                https://www.canada.ca/en/services.html   
13                https://www.canada.ca/en/services.html   
14                https://www.canada.ca/en/services.html   
15     

distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:49746 remote=tcp://127.0.0.1:45465>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:49748 remote=tcp://127.0.0.1:45465>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:34996 remote=tcp://127.0.0.1:41919>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:34998 remote=tcp://127.0.0.1:41919>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:49754 remote=tcp://127.0.0.1:45465>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:49756 remote=tcp://127.0.0.1:45465>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:41474 remote=tcp://127.0.0.1:32785>


In [20]:
# Get hs-analytics scripts
hs = df[df.script_url.str.contains('hs-analytics')]
hs.compute()

argument_0                       symbol  \
27753       None       window.document.cookie   
27754       None       window.document.cookie   
27755       None       window.document.cookie   
27756       None       window.document.cookie   
27757       None       window.document.cookie   
27758       None       window.document.cookie   
27759       None       window.document.cookie   
27760       None   window.navigator.userAgent   
27761       None       window.document.cookie   
27762       None       window.document.cookie   
27763       None       window.document.cookie   
27764       None       window.document.cookie   
27765       None       window.document.cookie   
27766       None       window.document.cookie   
27767       None       window.document.cookie   
27768       None       window.document.cookie   
27769       None       window.document.cookie   
27770       None       window.document.cookie   
27771       None       window.document.cookie   
27772       None       window.document.cookie   
27773       None       window.document.cookie   
27774       None       window.document.cookie   
27775       None       window.document.cookie   
27776       None       window.document.cookie   
27777       None       window.document.cookie   
27778       None     window.screen.colorDepth   
27779       None    window.navigator.language   
27780       None    window.navigator.language   
27781       None       window.document.cookie   
27782       None   window.navigator.userAgent   
...          ...                          ...   
20348       None       window.document.cookie   
20349       None       window.document.cookie   
20350       None       window.document.cookie   
20351       None       window.document.cookie   
20352       None       window.document.cookie   
20353       None       window.document.cookie   
20354       None       window.document.cookie   
20355       None       window.document.cookie   
20356       None       window.document.cookie   
20357       None       window.document.cookie   
20358       None       window.document.cookie   
20359       None       window.document.cookie   
20360       None       window.document.cookie   
20361       None       window.document.cookie   
20362       None       window.document.cookie   
20363       None       window.document.cookie   
20364       None     window.screen.colorDepth   
20365       None    window.navigator.language   
20366       None    window.navigator.language   
20367       None       window.document.cookie   
20368       None   window.navigator.userAgent   
20369       None    window.navigator.language   
20370       None     window.screen.colorDepth   
20371       None        window.sessionStorage   
20372       None          window.localStorage   
20373       None    window.navigator.platform   
20374       None  window.navigator.doNotTrack   
20375       None       window.document.cookie   
20376       None       window.document.cookie   
20377       None       window.document.cookie   

                                           location  \
27753  https://www.accenture.com/ca-en/terms-of-use   
27754  https://www.accenture.com/ca-en/terms-of-use   
27755  https://www.accenture.com/ca-en/terms-of-use   
27756  https://www.accenture.com/ca-en/terms-of-use   
27757  https://www.accenture.com/ca-en/terms-of-use   
27758  https://www.accenture.com/ca-en/terms-of-use   
27759  https://www.accenture.com/ca-en/terms-of-use   
27760  https://www.accenture.com/ca-en/terms-of-use   
27761  https://www.accenture.com/ca-en/terms-of-use   
27762  https://www.accenture.com/ca-en/terms-of-use   
27763  https://www.accenture.com/ca-en/terms-of-use   
27764  https://www.accenture.com/ca-en/terms-of-use   
27765  https://www.accenture.com/ca-en/terms-of-use   
27766  https://www.accenture.com/ca-en/terms-of-use   
27767  https://www.accenture.com/ca-en/terms-of-use   
27768  https://www.accenture.com/ca-en/terms-of-use   
27769  https://www.accenture.com

In [38]:
total=11292867
hs_analytics=13848
Percentage= ((hs_analytics/total)*100)
Percentage

0.12262607892220814

In [21]:
# Get akam scripts
akam = df[df.script_url.str.contains('/akam/')]
akam.compute()

argument_0  \
6091                    None   
6092                    None   
6093                    None   
6094                    None   
6095                    None   
6096                    None   
6097                    None   
6098                    None   
6099                    None   
6100                    None   
6101                    None   
6102                    None   
6103                    None   
6104                    None   
6105                    None   
6106                    None   
6107                    None   
6108                    None   
6109   __akfp_storage_test__   
6110   __akfp_storage_test__   
6111                    None   
6112   __akfp_storage_test__   
6113   __akfp_storage_test__   
6114                      2d   
6115                    None   
6116                    None   
6117                    None   
6118   Soft Ruddy Foothold 2   
6119                    None   
6120                     100   
...                      ...   
37135                   None   
37136                   None   
37137                   None   
37138                   None   
37139                   None   
37140                   None   
37141                   None   
37142                   None   
37143                   None   
37144                   None   
37145                   None   
37146                   None   
37147                   None   
37148                   None   
37149                   None   
37150                   None   
37151                   None   
37152                   None   
37153                   None   
37154                   None   
37155                   None   
37156                   None   
37157                   None   
37158                   None   
37159                   None   
37160                   None   
37161                   None   
37162                   None   
37163                     2d   
37171                      0   

                                                  symbol  \
6091    window.navigator.plugins[Shockwave Flash].length   
6092      window.navigator.plugins[Shockwave Flash].name   
6093   window.navigator.plugins[Shockwave Flash].desc...   
6094   window.navigator.plugins[Shockwave Flash].file...   
6095   window.navigator.mimeTypes[application/x-shock...   
6096   window.navigator.mimeTypes[application/x-shock...   
6097   window.navigator.mimeTypes[application/x-shock...   
6098    window.navigator.plugins[Shockwave Flash].length   
6099      window.navigator.plugins[Shockwave Flash].name   
6100   window.navigator.plugins[Shockwave Flash].desc...   
6101   window.navigator.plugins[Shockwave Flash].file...   
6102   window.navigator.mimeTypes[application/futures...   
6103   window.navigator.mimeTypes[application/futures...   
6104   window.navigator.mimeTypes[application/futures...   
6105    window.navigator.plugins[Shockwave Flash].length   
6106                            window.screen.colorDepth   
6107                            window.screen.pixelDepth   
6108                                 window.localStorage   
6109                              window.Storage.setItem   
6110                           window.Storage.removeItem   
6111                               window.sessionStorage   
6112                              window.Storage.setItem   
6113                           window.Storage.removeItem   
6114                        HTMLCanvasElement.getContext   
6115                  CanvasRenderingContext2D.fillStyle   
6116                       CanvasRenderingContext2D.font   
6117               CanvasRenderingContext2D.textBaseline   
6118                   CanvasRenderingContext2D.fillText   
6119                  CanvasRenderingContext2D.fillStyle   
6120                   CanvasRenderingContext2D.fillRect   
...                                                  ...   
37135                       window.navigator.appCodeName   
37136                        window.navigator.appVers

In [39]:
total=11292867
akamscripts=60975 
Percentage= ((akamscripts/total)*100)
Percentage

0.5399426027066466

In [22]:
# Get fingerprintjs2 scripts
fp2 = df[df.script_url.str.contains('fingerprint2')]
fp2.compute()

argument_0  \
7703                                None   
7704                                None   
7705                                None   
7706                                None   
7707                                None   
7708                                None   
7709                                None   
7710                                  2d   
7711                                None   
7712                                None   
7713                                None   
7714                                  2d   
7715                                   0   
7716                                   2   
7717                                   5   
7718                                None   
7719                                None   
7720                                 125   
7721                                None   
7722                                None   
7723   Cwm fjordbank glyphs vext quiz, 😃   
7724                                None   
7725                                None   
7726   Cwm fjordbank glyphs vext quiz, 😃   
7727                                None   
7728                                None   
7729                                  50   
7730                                None   
7731                                None   
7732                                 100   
...                                  ...   
34005                               None   
34006                               None   
34007  Cwm fjordbank glyphs vext quiz, 😃   
34008                               None   
34009                               None   
34010                                 50   
34011                               None   
34012                               None   
34013                                100   
34014                               None   
34015                               None   
34016                                 75   
34017                               None   
34018                               None   
34019                                 75   
34020                                 75   
34021                            evenodd   
34022                               None   
34023                                 2d   
34024                              webgl   
34025                              webgl   
34026                               None   
34027                               None   
34028                               None   
34029                               None   
34030                               None   
34031                               None   
34032                               None   
34033                               None   
34034                               None   

                                                  symbol  \
7703                          window.navigator.userAgent   
7704                           window.navigator.language   
7705                            window.screen.colorDepth   
7706                               window.sessionStorage   
7707                                 window.localStorage   
7708                           window.navigator.platform   
7709                           window.navigator.platform   
7710                        HTMLCanvasElement.getContext   
7711                             HTMLCanvasElement.width   
7712                            HTMLCanvasElement.height   
7713                             HTMLCanvasElement.style   
7714                        HTMLCanvasElement.getContext   
7715                       CanvasRenderingContext2D.rect   
7716                       CanvasRenderingContext2D.rect   
7717              CanvasRenderingContext2D.isPointInPath   
7718               CanvasRenderingContext2D.textBaseline   
7719                  CanvasRenderingContext2D.fillStyle   
7720                   CanvasRenderingContext2D.fillRect   
7721                  CanvasRenderingContext2D.fillStyle   
7722                       CanvasRenderingContext2D.font   
7723                   CanvasRenderingCon

In [40]:
total=11292867
fingerprintjs2=10649
Percentage= ((fingerprintjs2/total)*100)
Percentage

0.09429846291468766

Let's finish by looking for all the different ways fingerprintjs2 (or scripts very like it) may be referenced in the dataset.  Instead of just looking for fingerprintjs2 in the name we can look for the characteristic value `Cwm fjordbank glyphs vext quiz`

In [44]:
fp_arguments = df[df.argument_0.str.contains('Cwm fjordbank glyphs vext quiz')]
fp_arguments.compute()

argument_0  \
543       Cwm fjordbank glyphs vext quiz, 😃   
546       Cwm fjordbank glyphs vext quiz, 😃   
5106      Cwm fjordbank glyphs vext quiz, 😃   
5109      Cwm fjordbank glyphs vext quiz, 😃   
25438    Cwm fjordbank glyphs vext quiz, 馃槂   
25441    Cwm fjordbank glyphs vext quiz, 馃槂   
25493    Cwm fjordbank glyphs vext quiz, 馃槂   
25496    Cwm fjordbank glyphs vext quiz, 馃槂   
30331     Cwm fjordbank glyphs vext quiz, 😃   
30334     Cwm fjordbank glyphs vext quiz, 😃   
35402  Cwm fjordbank glyphs vext quiz, p   
35405  Cwm fjordbank glyphs vext quiz, p   
4931      Cwm fjordbank glyphs vext quiz, 😃   
4934      Cwm fjordbank glyphs vext quiz, 😃   
6371      Cwm fjordbank glyphs vext quiz, 😃   
6374      Cwm fjordbank glyphs vext quiz, 😃   
7723      Cwm fjordbank glyphs vext quiz, 😃   
7726      Cwm fjordbank glyphs vext quiz, 😃   
26006     Cwm fjordbank glyphs vext quiz, 😃   
26009     Cwm fjordbank glyphs vext quiz, 😃   
35703     Cwm fjordbank glyphs vext quiz, 😃   
35706     Cwm fjordbank glyphs vext quiz, 😃   
734       Cwm fjordbank glyphs vext quiz, 😃   
737       Cwm fjordbank glyphs vext quiz, 😃   
2385   Cwm fjordbank glyphs vext quiz, 😃   
2388   Cwm fjordbank glyphs vext quiz, 😃   
20598     Cwm fjordbank glyphs vext quiz, 😃   
20601     Cwm fjordbank glyphs vext quiz, 😃   
4344      Cwm fjordbank glyphs vext quiz, 😃   
4347      Cwm fjordbank glyphs vext quiz, 😃   
...                                     ...   
6951      Cwm fjordbank glyphs vext quiz, 😃   
6954      Cwm fjordbank glyphs vext quiz, 😃   
20655     Cwm fjordbank glyphs vext quiz, 😃   
20658     Cwm fjordbank glyphs vext quiz, 😃   
21214     Cwm fjordbank glyphs vext quiz, 😃   
21217     Cwm fjordbank glyphs vext quiz, 😃   
36439     Cwm fjordbank glyphs vext quiz, 😃   
36442     Cwm fjordbank glyphs vext quiz, 😃   
36605     Cwm fjordbank glyphs vext quiz, 😃   
36608     Cwm fjordbank glyphs vext quiz, 😃   
1319      Cwm fjordbank glyphs vext quiz, 😃   
1322      Cwm fjordbank glyphs vext quiz, 😃   
22383     Cwm fjordbank glyphs vext quiz, 😃   
22386     Cwm fjordbank glyphs vext quiz, 😃   
23064  Cwm fjordbank glyphs vext quiz, 😃   
23067  Cwm fjordbank glyphs vext quiz, 😃   
333          Cwm fjordbank glyphs vext quiz   
336          Cwm fjordbank glyphs vext quiz   
5008   Cwm fjordbank glyphs vext quiz, 😃   
5011   Cwm fjordbank glyphs vext quiz, 😃   
32574     Cwm fjordbank glyphs vext quiz, 😃   
32577     Cwm fjordbank glyphs vext quiz, 😃   
32819     Cwm fjordbank glyphs vext quiz, 😃   
32822     Cwm fjordbank glyphs vext quiz, 😃   
33917     Cwm fjordbank glyphs vext quiz, 😃   
33920     Cwm fjordbank glyphs vext quiz, 😃   
34004     Cwm fjordbank glyphs vext quiz, 😃   
34007     Cwm fjordbank glyphs vext quiz, 😃   
35161  Cwm fjordbank glyphs vext quiz, 😃   
35164  Cwm fjordbank glyphs vext quiz, 😃   

                                  symbol  \
543    CanvasRenderingContext2D.fillText   
546    CanvasRenderingContext2D.fillText   
5106   CanvasRenderingContext2D.fillText   
5109   CanvasRenderingContext2D.fillText   
25438  CanvasRenderingContext2D.fillText   
25441  CanvasRenderingContext2D.fillText   
25493  CanvasRenderingContext2D.fillText   
25496  CanvasRenderingContext2D.fillText   
30331  CanvasRenderingContext2D.fillText   
30334  CanvasRenderingContext2D.fillText   
35402  CanvasRenderingContext2D.fillText   
35405  CanvasRenderingContext2D.fillText   
4931   CanvasRenderingContext2D.fillText   
4934   CanvasRenderingContext2D.fillText   
6371   CanvasRenderingContext2D.fillText   
6374   CanvasRenderingContext2D.fillText   
7723   CanvasRenderingContext2D.fillText   
7726   CanvasRenderingContext2D.fillText   
26006  CanvasRenderingContext2D.fillText   
26009  CanvasRenderingContext2D.fillText   
35703  CanvasRenderingContext2D.fillText   
35706  CanvasRenderingContext2D.fillText   
734    CanvasRenderingContext2D.fillText   
737    CanvasRenderingContext2D.fillText   
2385   CanvasRende

In [45]:
tal=11292867
fp_script=10649
Percentage= ((fp_script/total)*100)
Percentage

0.09429846291468766

## Conclusion

It can be seen, out of 11292867 scripts  13848 contain hs_analytics, 60975 akam, 10649 fingerprintjs in the url which is 0.123%, 0.539%, 0.094% of the total scripts respectively.

Amongst the three akam scripts have the greatest contribution of 0.539%.

If we consider three of the scripts together it forms 0.756% of the total scripts.

